packages/compiler/src/source.ts

```diff
@@ -48,6 +32,74 @@ export async function fetchComponentSources(
       });
       return sources;
     },
-    {} as { [key: string]: BOSModule }
+    {} as ComponentSourcesResponse
+  );
+}
+
+function isNotABlockEntry<T>(entryKey: string, entryValue: any): entryValue is T {
+  return typeof entryValue !== 'number' && entryKey !== BLOCK_HEIGHT_KEY;
+}
+
+function prepareSourceWithBlockHeight(response: SocialComponentsByAuthorWithBlockHeight) {
+  return Object.entries(response).reduce(
+    (sources, [entryKey, entryValue]) => {
+      if (isNotABlockEntry<SocialComponentWithBlockHeight>(entryKey, entryValue)) {
+        const { [SOCIAL_COMPONENT_NAMESPACE]: component, [BLOCK_HEIGHT_KEY]: componentBlockHeight } = entryValue;
+
+        if (isNotABlockEntry<SocialWidgetWithBlockHeight>(SOCIAL_COMPONENT_NAMESPACE, component)) {
+
+          Object.entries(component).forEach(([componentKey, componentValue]) => {
+
+            if (isNotABlockEntry<ComponentEntryWithBlockHeight>(componentKey, componentValue)) {
+              const sourceKey = `${entryKey}/${componentKey}`;
+              sources[sourceKey] = { component: componentValue[""][""], css: componentValue.css[""], blockHeight: componentBlockHeight };
+            }
+
+          });
+        }
+      }
+
+      return sources;
+    },
+    {} as ComponentSourcesResponse
+  );
+}
+
+export async function fetchComponentSources(
+  social: SocialDb,
+  componentPaths: string[],
+  options = { withBlockHeight: true }
+): Promise<Record<string, BOSModule>> {
+  /*
+    Typically, you'd want to pass a generic to `social.get<MyType>()`. This generic
+    would be wrapped by DeepPartial (recursively flagging all properties as possibly
+    undefined). However, we want this function to actually throw an error if it's trying
+    to access a component (or property) that doesn't exist. That's why we cast with
+    `as SocialComponentsByAuthor` - which will retain our purposefully "dangerous"
+    `any` typings.
+  */
+
+  const keys = componentPaths.map(
+    (p) => p.split('/').join(`/${SOCIAL_COMPONENT_NAMESPACE}/`) + '/*'
   );
+
+  const socialGetParams: SocialGetParams = {
+    keys,
+  };
+
+  if (options?.withBlockHeight) {
+    const optionsWithBlockHeight = {
+      with_block_height: true,
+    };
+
+    socialGetParams.options = optionsWithBlockHeight;
+
+    const response = (await social.get(socialGetParams)) as SocialComponentsByAuthorWithBlockHeight;
```

Comment:
💄 `with_block_height` can be inlined into params
```js
(await social.get({ ...socialGetParams, with_block_height: true }))
```

Reply:
This param should be set on the `options` key, so it might looks like this:
```
      {
        ...socialGetParams, ...{
          options: { with_block_height: true, }
        }
      }
```
But I think it's cleaner to keep the options in a separate variable. 
I guess we could improve the naming to be `socialOptions` instead of `optionsWithBlockHeight`.
@andy-haynes would you be ok with that?

Reply:
I think it would be a more natural fit if we had other configurable options to pass along, but for our straightforward use case it just seems a little superfluous. Just an observation 🙂